File: ta_lib/strategies/signal/src/signalline/dso_signalline.rs
use base::prelude::*;
use core::prelude::*;
use momentum::dso;

pub struct DsoSignalLineSignal {
    smooth_type: Smooth,
    smooth_period: usize,
    k_period: usize,
    d_period: usize,
}

impl DsoSignalLineSignal {
    pub fn new(smooth_type: Smooth, smooth_period: f32, k_period: f32, d_period: f32) -> Self {
        Self {
            smooth_type,
            smooth_period: smooth_period as usize,
            k_period: k_period as usize,
            d_period: d_period as usize,
        }
    }
}

impl Signal for DsoSignalLineSignal {
    fn lookback(&self) -> usize {
        let period = std::cmp::max(self.smooth_period, self.k_period);
        std::cmp::max(period, self.d_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let (k, d) = dso(
            &data.close(),
            self.smooth_type,
            self.smooth_period,
            self.k_period,
            self.d_period,
        );

        (k.cross_over(&d), k.cross_under(&d))
    }
}

File: strategy/generator/signal/signalline/dso_signalline.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.smooth import Smooth
from core.models.source import SourceType
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class DsoSignalLineSignal(Signal):
    type: Signal = SignalType.DsoSignalLine
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    smooth_type: Parameter = StaticParameter(Smooth.EMA)
    smooth_period: Parameter = StaticParameter(13.0)
    k_period: Parameter = StaticParameter(8.0)
    d_period: Parameter = StaticParameter(9.0)
